auth: do not create default 'cassandra:cassandra' superuser

DarioMirovic opened the pull request:
### PR goal

This patch series removes creation of default 'cassandra:cassandra' superuser on system start.

### The motivation

Disable creation of a superuser with default 'cassandra:cassandra' credentials to improve security. The current flow requires clients to create another superuser and then drop the default `cassandra:cassandra' role. For those who do, there is a time window where the default credentials exist. For those who do not, that role stays. We want to improve security by forcing the client to either use config to specify default values for default superuser name and password or use cqlsh over maintenance socket connection to explicitly create/alter a superuser role.

The patch series:
- Enable role modification over the maintenance socket
- Stop using default 'cassandra' value for default superuser, skipping creation instead

Design document: https://scylladb.atlassian.net/wiki/spaces/RND/pages/165773327/Drop+default+cassandra+superuser

Fixes scylladb/scylla-enterprise#5657

This is an improvement. It does not need a backport.

mykaul commented:
Wow. I couldn't believe it passed CI, but then I saw you do create this user for the tests. OK...

CC @fruch  for awareness of this change.

nyh commented:
> Wow. I couldn't believe it passed CI, but then I saw you do create this user for the tests. OK...

Yes, I'll also need to fix test/cqlpy/run.py, but that will be easy and it's not needed for the CI.

But How does dtest, which you didn't fix, work with this change?

DarioMirovic commented:
Retriggering the CI. The failure seems to be caused by the CI, not by the PC changes.
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/22196
```
21:15:01  Found a total of 15 nodes with the 'builders_x86' label
21:15:01  [Pipeline] node
21:15:15  Still waiting to schedule task
21:15:15  Waiting for next available executor on ‘[builders_x86](https://jenkins.scylladb.com/label/builders_x86/)’
21:15:15  Still waiting to schedule task
21:15:15  Waiting for next available executor on ‘[builders_x86](https://jenkins.scylladb.com/label/builders_x86/)’
21:15:16  Still waiting to schedule task
21:15:16  Waiting for next available executor on ‘[builders_x86](https://jenkins.scylladb.com/label/builders_x86/)’
23:15:12  Agent i-0dbbb0612ddf59da2 was deleted; cancelling node body
23:15:12  [Pipeline] // node
23:15:12  [Pipeline] // node
23:15:12  [Pipeline] // node
23:15:13  [Pipeline] }
23:15:13  Failed in branch dev
23:15:13  [Pipeline] }
23:15:13  Failed in branch debug
23:15:13  [Pipeline] }
23:15:13  Failed in branch release
23:15:14  [Pipeline] // parallel
23:15:14  Did you forget the `def` keyword? WorkflowScript seems to be setting a field named metadataFiles (to a value of type ArrayList) which could lead to memory leaks or other issues.
...
23:15:16  Finished: FAILURE
```
Also, on the checks panel on the PR it says "Waiting for status to be reported - Waiting for builder", and a yellow/orange dot, not failure report and red dot. This looks irregular.

nyh commented on auth/password_authenticator.cc:
```diff
@@ -54,11 +54,18 @@ static const class_registrator<
+
+    auto superuser_name = get_config_value(cfg.auth_superuser_name());
+    if (superuser_name) {
+        return *superuser_name;
```
nitpick: you can return superuser_name instead of *superuser_name.

DarioMirovic replied:
Reverted from optional<string> to string, comment not relevant now

nyh commented on auth/password_authenticator.cc:
```diff
@@ -125,12 +132,13 @@ future<> password_authenticator::migrate_legacy_metadata() const {
 
 future<> password_authenticator::legacy_create_default_if_missing() {
-    const auto exists = co_await legacy::default_role_row_satisfies(_qp, &has_salted_hash, _superuser);
+    SCYLLA_ASSERT(_superuser);
```
In general, I'm not happy about adding SCYLLA_ASSERT (https://github.com/scylladb/scylladb/issues/7871). In this case, why is _superuser even an "optional" if you plan to assert it must be set?
Or, maybe you'll remove this assert in a a following patch? I'll see...

nyh commented on main.cc:
```diff
@@ -2114,6 +2114,13 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
 
+                // Notify maintenance auth service that maintenance mode is starting
+                maintenance_auth_service.invoke_on_all([](auth::service& svc) {
+                    auto& rm = dynamic_cast<auth::maintenance_socket_role_manager&>(svc.underlying_role_manager());
```
Are we sure this won't crash because the type of wrong?

DarioMirovic replied:
`maintenance_auth_service` is always initialized with `maintenance_socket_role_manager`.

main.cc:
```cpp
                maintenance_auth_config.role_manager_java_name = sstring{auth::maintenance_socket_role_manager_name};

                maintenance_auth_service.start(perm_cache_config, std::ref(qp), std::ref(group0_client),  std::ref(mm_notifier), std::ref(mm), maintenance_auth_config, maintenance_socket_enabled::yes, std::ref(auth_cache), std::ref(hashing_worker)).get();
```

auth/service.cc in `auth::service::service`:
```cpp
                      create_object<role_manager>(sc.role_manager_java_name, qp, g0, mm, cache),
```

This seemed like the least intrusive way for this. We could, of course, have a virtual function or something else, but having all role managers with `virtual future<> enable_role_operations();` felt a bit confusing.

nyh commented on service/client_state.cc:
```diff
@@ -56,11 +56,18 @@ void service::client_state::validate_login() const {
 }
 
+bool service::client_state::is_from_maintenance_socket() const {
+    return _auth_service->is_used_by_maintenance_socket();
```
I don't understand what I'm seeing here:
Please correct me if I'm wrong, but auth_service has "is_used_by_maintenance_socket" if **also** used through maintenance socket. So how can we assume that some other socket's request is "from maintenance socket" and therefore "not anonymous"?
Unless I'm misunderstanding how this works, this looks incorrect to me.

If the "auth service" used by the maintenance socket is a separate "auth service" used only by it, please add a comment to is_used_by_maintenance_socket() clarifying this situation - that it means it is used **only** by a maintenance socket. Maybe it should even be named accordingly is_maintenance_socket() or for_maintenance_socket() - emphasizing that this object is only for the maintenance socket and nothing else.

DarioMirovic commented:
> > > When I asked about dtest I didn't mean the pieces of dtest we copied into test/cluster, I mean the separate dtest repository.
> > 
> > 
> > Right. That part is not done yet.
> 
> So we expect the dtest part of this CI to fail until [scylladb/scylla-dtest#6457](https://github.com/scylladb/scylla-dtest/pull/6457) gets in?

Exactly. This CI uses some of the scylla-dtest tests, and they fail on authentication. I already tested locally dtest tests that failed, just with injected credentials for `cassandra:cassandra` in the config, and the tests passed. I also have to check SCTs config before this PR gets merged.

DarioMirovic commented on docs/operating-scylla/security/create-superuser.rst:
```diff
@@ -1,22 +1,108 @@
 
-To improve security, we recommend creating a custom superuser. You should:
+There are three ways you can create a superuser in ScyllaDB:
+- :ref:`Using an existing superuser account to create a new superuser role <create-superuser-using-existing-superuser>`
```
Blank line before bullet points is mandatory it seems. Updated.

nuivall commented on auth/standard_role_manager.cc:
```diff
@@ -176,6 +176,9 @@ future<> standard_role_manager::create_legacy_metadata_tables_if_missing() const
 }
 
 future<> standard_role_manager::legacy_create_default_role_if_missing() {
+    if (_superuser.empty()) {
```
why this? I think we shouldn't change legacy flow

nuivall replied:
the only problem is that it's not equivalent to ASSERT, throwing this inside do_after_system_ready will only cause loop for 1 minute and debug info log

but we can leave as is, v1 is going to be deleted anyway

nuivall commented:
`auth: make _superuser field optional` - I don't understand this commit nor its description. It seems in some cases it makes the field mandatory. And also it stays `std::string`, not `optional<std::string>`. It would be good to describe in commit message something around before and after state  for such a change. Also it says "prepares" but doesn't explain why.

nuivall commented on auth/password_authenticator.cc:
```diff
@@ -53,11 +53,15 @@ static const class_registrator<
-std::string password_authenticator::default_superuser(const db::config& cfg) {
-    return std::string(get_config_value(cfg.auth_superuser_name(), DEFAULT_USER_NAME));
+std::string password_authenticator::default_superuser(cql3::query_processor& qp) {
+    const auto& cfg = qp.db().get_config();
```
why this change?

DarioMirovic replied:
In a later patch, when default superuser creation for non-legacy mode is created, we have to use `legacy_mode(qp)` to check if we are in the legacy mode. I can either make the change here or in that final commit.

nuivall replied:
ok, better to make the change where it's being used

nuivall commented on auth/password_authenticator.cc:
```diff
@@ -178,12 +188,12 @@ future<> password_authenticator::maybe_create_default_password() {
     }
     const auto update_query = update_row_query();
-    co_await collect_mutations(_qp, batch, update_query, {salted_pwd, _superuser});
+    co_await collect_mutations(_qp, batch, update_query, {std::move(salted_pwd), _superuser});
```
same as https://github.com/scylladb/scylladb/pull/27215/changes/bf9aca5533e785f2840f30f8d56a1afdf36cf496#r2650972352

DarioMirovic replied:
Reverted the change

nuivall commented:
```
auth: maintenance socket connection access rights and superuser status

Maintenance auth service uses `allow_all_authenticator`. To allow
role modification statements over the maintenance socket connections,
we need to treat the maintenance socket connections as superusers and
give them proper access rights.
```
Not clear here, why we can't use the same code? maintenance_socket_role_manager::is_superuser already returns unconditionally true.

nuivall reviewed (changes requested):
left some comments

DarioMirovic commented on auth/common.hh:
```diff
@@ -46,6 +46,8 @@ extern constinit const std::string_view USERS_CF;
 constexpr std::string_view DEFAULT_SUPERUSER_NAME("cassandra");
 extern constinit const std::string_view AUTH_PACKAGE_NAME;
 
+constexpr std::string_view MAINTENANCE_ROLE_NAME("$MAINTENANCE$");
```
At the moment there is nothing to prevent somebody from creating a role with this name. It can be done by connecting via tcp socket. Only the maintenance socket uses maintenance socket role manager, which forbids this. It is something that could and should be improved. The first iteration has it mentioned in the documentation, but having it explicitly handled by the code is a safer solution.

nuivall commented:
We have some default superuser name code in `service::describe_roles`,
```
        // Even after applying formatting to a role, `formatted_role_name` can only equal `meta::DEFAULT_SUPER_NAME`
        // if the original identifier was equal to it.
        const sstring role_part = formatted_role_name == meta::DEFAULT_SUPERUSER_NAME
                ? seastar::format("IF NOT EXISTS {}", formatted_role_name)
                : formatted_role_name;
```

I wonder if this needs any adjustments with your changes.

nuivall commented:
Please improve name of the commit: `auth: maintenance auth service configuration update` - it only changes how objects are created, doesn't change the configuration itself. Also good practice is to put verb at the beginning of commit message.

nuivall commented:
```
This patch lets `maintenance_socket_role_manager` know if it works under
maintenance mode and returns appropriate error message when role
operations execution is requested.
```
Is this only to provide different errors? In maintenence mode they will probably fail due to missing quorum or something like that. Maybe that's actually providing better or equally good info and we don't need this extra complexity?

DarioMirovic commented:
In V4:
- Removed `$MAINTENANCE$` named user for maintenance socket connection
    - Handled it with flags, but not inside auth module, to avoid mutual dependency
- Added `_bypass_auth_checks` in `client_state`, and `has_superuser` method wrapper used around `auth::has_superuser`
- Removed class registrator and `service_config` usage in auth module
- Set maintenance socket path default from `ignore` to `workdir`

cc @nuivall

ptrsmrn commented on auth/allow_all_authenticator.cc:
```diff
@@ -9,19 +9,9 @@
 #include "auth/allow_all_authenticator.hh"
```
Yes.

DarioMirovic commented on test/cluster/auth_cluster/test_maintenance_socket.py:
```diff
@@ -67,3 +70,124 @@ async def test_maintenance_socket(manager: ManagerClient):
+    """
+    Test that when no default superuser is configured:
+    1. No 'cassandra' user exists.
+    2. We can manage roles via the maintenance socket.
```
I split the new test into two. One tests for no default superuser. The other tests various role operations. I did not want to split it further, I think it would be an overkill. You can check it out, I pushed the code.

nuivall commented on auth/service.cc:
```diff
@@ -941,4 +949,115 @@ future<> migrate_to_auth_v2(db::system_keyspace& sys_ks, ::service::raft_group0_
+
+namespace {
+
+// Helper to check if a name matches any of the given suffixes (case-insensitive comparison for the suffix part)
```
such matching seems to be an overkill here, you can just create a view on name with substr starting from last "." character and match with ==

nuivall commented:
`auth: maintenance_socket_role_manager::_is_maintenance_mode` -> I think I commented something similar here before but perhaps it got lost. Please change to meaningful commit message.

nuivall commented:
`auth: maintenance socket connection access rights and superuser status` - this commit title is missing a verb, it's unclear what we do here, add access right status or remove them or something else?

nuivall commented on docs/operating-scylla/security/create-superuser.rst:
```diff
@@ -1,111 +1,108 @@
+------------------------------------------------------
 
-   .. code::
+If no superuser account exists in the cluster, you can create a superuser using the ScyllaDB Maintenance Socket.
```
This option should be mentioned as the 1st one. Also in the initial note it we could direct to this method for new clusters

nuivall commented:
@annastuchlik could you review documentation changes? Especially that it will affect all new clusters which want to have auth enabled.

nuivall commented on test/cluster/auth_cluster/test_maintenance_socket.py:
```diff
@@ -67,3 +70,138 @@ async def test_maintenance_socket(manager: ManagerClient):
+    server = await manager.server_add(config=config, start=False)
+
+    # Remove the default superuser configuration to simulate a clean start without seeding
+    await manager.server_remove_config_option(server.server_id, 'auth_superuser_name')
```
wouldn't setting it to empty in `config = {` above work?

annastuchlik commented on docs/operating-scylla/security/create-superuser.rst:
```diff
@@ -1,111 +1,108 @@
+.. _create-superuser-in-config-file:
 
-    cqlsh -u <custom_superuser name> -p <custom_superuser_password>
+Setting Custom Superuser Credentials in scylla.yaml
```
The same question here. As there's no default superuser now, perhaps we should remove "custom" from all the places?
"Custom" suggest there's a default.

DarioMirovic commented:
> ```
> auth: maintenance socket connection access rights and superuser status
> 
> Maintenance auth service uses `allow_all_authenticator`. To allow
> role modification statements over the maintenance socket connections,
> we need to treat the maintenance socket connections as superusers and
> give them proper access rights.
> ```
> 
> Not clear here, why we can't use the same code? maintenance_socket_role_manager::is_superuser already returns unconditionally true.

@nuivall I updated the commit message. It should be clearer now what happens and why in this patch.

nuivall commented on service/qos/service_level_controller.hh:
```diff
@@ -293,19 +293,11 @@ public:
-        // Special case:
-        // -------------
-        // The maintenance socket can communicate with Scylla before `auth_integration`
-        // is registered, and we need to prepare for it.
```
the 
```
        // The maintenance socket can communicate with Scylla before `auth_integration`
        // is registered, and we need to prepare for it.
```

part is actually important info, why deleting it?

nuivall commented on service/qos/service_level_controller.cc:
```diff
@@ -661,8 +661,11 @@ future<scheduling_group> service_level_controller::get_user_scheduling_group(con
 scheduling_group service_level_controller::get_cached_user_scheduling_group(const std::optional<auth::authenticated_user>& usr) {
-    SCYLLA_ASSERT(_auth_integration != nullptr);
-    return _auth_integration->get_user_cached_scheduling_group(usr);
+    if (_auth_integration == nullptr) {
```
please add a comment, the same as in 1st commit

nuivall commented on cql3/statements/describe_statement.cc:
```diff
@@ -659,8 +659,7 @@ future<std::vector<std::vector<managed_bytes_opt>>> schema_describe_statement::d
             if (config.with_hashed_passwords) {
-                const auto maybe_user = client_state.user();
-                if (!maybe_user || !co_await auth::has_superuser(auth_service, *maybe_user)) {
+                if (!co_await client_state.has_superuser()) {
```
before there was an optional check here, now auth::has_superuser(*_auth_service, *_user) doesn't perform such check, why it's ok to skip?

nuivall commented on auth/maintenance_socket_role_manager.cc:
```diff
@@ -13,12 +13,33 @@
+            "role manager: enabling role operations not allowed in maintenance mode"));
+    }
+
+    if (_std_mgr.has_value()) {
```
that'd mean ensure_role_operations_are_enabled would be called twice? Probably also not allowed, but if allowed this would return wrong result. I think for simplification we can disallow here

nuivall commented on docs/operating-scylla/security/create-superuser.rst:
```diff
@@ -1,111 +1,95 @@
 any CQL command on the database resources.
 
-To improve security, we recommend creating a custom superuser. You should:
+There are three ways you can create a superuser in ScyllaDB:
```
three -> mentions only 2

nuivall commented on test/pylib/scylla_cluster.py:
```diff
@@ -142,6 +142,10 @@ def make_scylla_conf(mode: str, workdir: pathlib.Path, host_addr: str, seed_addr
 
         'maintenance_socket': socket_path,
 
+        'auth_superuser_name': 'cassandra',
```
shouldn't be set only for test/cluster/auth_cluster?

DarioMirovic commented on test/cluster/auth_cluster/test_maintenance_socket.py:
```diff
@@ -77,3 +78,162 @@ async def test_maintenance_socket(manager: ManagerClient):
+    rows = list(session.execute("SELECT role, is_superuser FROM system.roles"))
+    assert len(rows) == 0, f"Expected no roles, found: {rows}"
+
+    logger.info("Waiting for role manager to be ready")
```
Maintenance socket role manager is ready after the node joins the cluster. Maintenance auth service and maintenance socket role manager with it are instantiated before the node joins the cluster. It is needed because we also need maintenance socket connection in maintenance mode.

Startup order - maintenance mode:
- Maintenance auth starts
    - Maintenance auth service starts fully
    - Maintenance socket role manager starts in mode that does not allow role operations
- Maintenance mode starts
    - Now we can use maintenance auth service to connect to the maintenance socket and execute some operations, but no role operations

Startup order - normal mode:
- Maintenance auth starts
    - Maintenance auth service starts fully
    - Maintenance socket role manager starts in mode that does not allow role operations
- The node joins the cluster
- Maintenance socket role manager operations are enabled

Because of the time span we need for maintenance socket role manager to not be prepared to execute role operations, what I will do is create a helper function in the test that returns the maintenance session only after role manager is ready.

DarioMirovic commented on test/cluster/auth_cluster/test_maintenance_socket.py:
```diff
@@ -77,3 +78,162 @@ async def test_maintenance_socket(manager: ManagerClient):
+            c.shutdown()
+            return None
+
+    admin_session = await wait_for(try_admin_connect, time.time() + 60)
```
I noticed that the tests are significantly flaky if there is no retry. I get around 2-8 out of 100 failure rate. `cluster_con`, which creates `Cluster`, and `Cluster.connect` do not seem to be reliable by themselves. I have just retried without the loop, and multiple runs failed with `ConnectionRefusedError`. The port seemed to not be ready yet. With the loops the test passed each time in maybe 500-700 runs I tried.

DarioMirovic commented on test/cluster/auth_cluster/test_maintenance_socket.py:
```diff
@@ -77,3 +78,162 @@ async def test_maintenance_socket(manager: ManagerClient):
+        try:
+            session.execute("SELECT role FROM system.roles LIMIT 1")
+            return session
+        except InvalidRequest:
```
Combined handling of all failures under one `except` block

DarioMirovic commented on test/cluster/auth_cluster/test_maintenance_socket.py:
```diff
@@ -77,3 +78,162 @@ async def test_maintenance_socket(manager: ManagerClient):
+    """
+    config = {
+        **auth_config,
+        "authenticator": "PasswordAuthenticator",
```
Removed

DarioMirovic commented on auth/maintenance_socket_role_manager.cc:
```diff
@@ -64,110 +90,170 @@ future<T> operation_not_available_in_maintenance_mode_exception(std::string_view
 }
 
+template<typename T = void>
+future<T> operation_not_ready_exception(std::string_view operation) {
```
Renamed

DarioMirovic commented on auth/maintenance_socket_role_manager.cc:
```diff
@@ -13,12 +13,33 @@
+
+future<> maintenance_socket_role_manager::ensure_role_operations_are_enabled() {
+    if (_is_maintenance_mode) {
+        return make_exception_future<>(std::runtime_error(
```
Switched to `on_internal_error`. It is supposed to be used for internal errors, where the logic does not depend on external input. This is such a case.

DarioMirovic commented on auth/maintenance_socket_authenticator.hh:
```diff
@@ -0,0 +1,43 @@
+#include "db/consistency_level_type.hh"
+#include "auth/authenticator.hh"
+#include "auth/passwords.hh"
+#include "service/raft/raft_group0_client.hh"
```
Removed

DarioMirovic commented on auth/maintenance_socket_role_manager.cc:
```diff
@@ -90,23 +135,38 @@ future<bool> maintenance_socket_role_manager::can_login(std::string_view role_na
 }
 
-future<std::optional<sstring>> maintenance_socket_role_manager::get_attribute(std::string_view role_name, std::string_view attribute_name, ::service::query_state&) {
+future<std::optional<sstring>> maintenance_socket_role_manager::get_attribute(std::string_view role_name, std::string_view attribute_name, ::service::query_state& qs) {
```
I tried to increase patches that much, there are 18 already, and this change seems benign to add.

DarioMirovic commented on auth/maintenance_socket_role_manager.cc:
```diff
@@ -13,20 +13,10 @@
-        ::service::migration_manager&,
-        cache&> registration(sstring{maintenance_socket_role_manager_name});
-
+static constexpr std::string_view maintenance_socket_role_manager_name = "com.scylladb.auth.MaintenanceSocketRoleManager";
```
Updated, only needed name variables left, others moved inside `qualified_java_name` methods.

DarioMirovic commented on service/qos/service_level_controller.cc:
```diff
@@ -653,19 +653,11 @@ scheduling_group service_level_controller::auth_integration::get_user_cached_sch
-    if (!usr.has_value() || auth::is_anonymous(usr.value())) {
+    if (_auth_integration == nullptr) {
         return make_ready_future<scheduling_group>(get_default_scheduling_group());
+    } else {
```
Reverted, now using the suggested approach

nuivall reviewed (approved):
LGTM, only minor nits which may be left as is. Remaining https://github.com/scylladb/scylladb/pull/27215#issuecomment-3983034808 may be addressed in a follow-up, if work is necessary there.

nyh commented:
Thanks @DarioMirovic. I confirmed that test/cqlpy/run and test/alternator/run (for example) do work fine after this patch.
I guess the magic (for the case of test/cqlpy/run) is in test/cqlpy/run.py, which has:
```
        '--auth-superuser-name', 'cassandra',
        # password is "cassandra"
        '--auth-superuser-salted-password', '$6$x7IFjiX5VCpvNiFk$2IfjTvSyGL7zerp
V.wbY7mJjaRCrJ/68dtT3UpT.sSmNYz1bPjtn3mH.kJKFvaZ2T4SbVeBijjmwGjcb83LlV/',
```

Which creates the cassandra/cassandra superuser, even if it's no longer the default.

This is the good news. Like the old western, there's also the bad news and the ugly news:

The (little bit) bad news is that it broke "test/cqlpy/run --release" with some old versions (e.g., 5.2), which now fail on "error: unrecognised option '--auth-superuser-name'". I can easily fix that, and will send a patch (and I full well expected that there is a cost in maintaining this "--release" trick, and I am happy to pay that cost by maintaining it).

The ugly news is that you can no longer run these tests against a Scylla that you run manually, unless you remember to use these new options in your manual test run, as well as have a "--workdir" because "--maintenance-socket" now defaults to workdir. In fact I tried now to run tests against a Scylla I ran manually, and I couldn't get it to work - I'll continue to working on that tomorrow :-(

This isn't a deal-breaker - but it makes it even more difficult, as if it wasn't difficult enough already, to run Scylla manually without some long script to do it for you. But I guess it isn't really realistic any more to expect to remember how to run Scylla manually. Which is, frankly, sad. And didn't have to be this way. But that's the way it is...
